llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp

```diff
@@ -3661,6 +3661,53 @@ static bool hasAffectedValue(Value *V, SmallPtrSetImpl<Value *> &Affected,
   return false;
 }
 
+// This transformation enables the possibility of transforming fcmp + sel into
+// a fmaxnum/fminnum intrinsic.
+static Value *foldSelectIntoAddConstant(SelectInst &SI,
+                                        InstCombiner::BuilderTy &Builder) {
+  // Do this transformation only when select instruction gives NaN and NSZ
+  // guarantee.
+  auto *SIFOp = dyn_cast<FPMathOperator>(&SI);
+  if (!SIFOp || !SIFOp->hasNoSignedZeros() || !SIFOp->hasNoNaNs())
+    return nullptr;
+
+  // select((fcmp Pred, X, 0), (fadd X, C), C)
+  //      => fadd((select (fcmp Pred, X, 0), X, 0), C)
+  //
+  // Pred := OGT, OGE, OLT, OLE, UGT, UGE, ULT, and ULE
+  Instruction *FAdd;
+  Constant *C;
+  Value *X, *Z;
+  CmpInst::Predicate Pred;
+
+  // Note: OneUse check for `Cmp` is necessary because it makes sure that other
+  // InstCombine folds don't undo this transformation and cause an infinite
+  // loop. Furthermore, it could also increase the operation count.
+  if (match(&SI, m_Select(m_OneUse(m_FCmp(Pred, m_Value(X), m_Value(Z))),
+                          m_OneUse(m_Instruction(FAdd)), m_Constant(C))) ||
+      match(&SI, m_Select(m_OneUse(m_FCmp(Pred, m_Value(X), m_Value(Z))),
+                          m_Constant(C), m_OneUse(m_Instruction(FAdd))))) {
+    if (!match(Z, m_AnyZeroFP()))
+      return nullptr;
+
+    // Only these relational predicates can be transformed into maxnum/minnum
+    // intrinsic.
+    if (!CmpInst::isRelational(Pred))
+      return nullptr;
+
+    if (!match(FAdd, m_FAdd(m_Specific(X), m_Specific(C))))
+      return nullptr;
+
+    Value *NewSelect =
+        Builder.CreateSelect(SI.getCondition(), X, Z, SI.getName(), &SI);
+    cast<Instruction>(NewSelect)->setFastMathFlags(SI.getFastMathFlags());
+
+    return Builder.CreateFAddFMF(NewSelect, C, FAdd, FAdd->getName());
```

Comment:
I don't think you can just pass through the flags from the original fadd. I believe you need to and with the fcmp's flags

Reply:
OK, playing with the above alive link, I think this wants the union of the value flags, and the intersection of the rewrite flags. We ought to have a helper directly in FMF for this case, since this is a common pattern.

So or of nsz / nnan/ ninf, and of reassoc, afn, contract

Reply:
Sorry, I didn’t quite understand why we need to consider fcmp flags. Could you please elaborate?

Reply:
I see, but without this, I cannot create a failing test case. Do you have any broken scenarios in mind?

Reply:
Fixed rewrite-based flags.